bump Scala bridge to use 2.13.0-RC1

eed3si9n opened the pull request:
This also points all 2.13.x and above to use 2.13.0-RC1 compatible source JAR as the compiler bridge. We are fairly certain we can maintain source-compatibility of compiler bridge from 2.13.0-RC to 2.13.x.

See also https://github.com/sbt/zinc/issues/78 :)

eed3si9n committed:
bump Scala bridge to use 2.13.0-RC1

We are fairly certain we can maintain source-compatibility of compiler bridge from 2.13.0-RC to 2.13.x.

See also https://github.com/sbt/zinc/issues/78 :)

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=develop&zinc=pull/651/head&io=develop&librarymanagement=develop&util=develop&website=develop&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

dwijnand commented on internal/zinc-ivy-integration/src/main/scala/sbt/internal/inc/ZincComponentCompiler.scala:
```diff
@@ -40,14 +40,12 @@ private[sbt] object ZincComponentCompiler {
-        case sc if (sc startsWith "2.12.")       => "compiler-bridge_2.12"
-        case "2.13.0-M1"                         => "compiler-bridge_2.12"
-        case sc if (sc startsWith "2.13.0-pre-") => "compiler-bridge_2.13.0-M5"
-        case sc if (sc startsWith "2.13.0-M")    => "compiler-bridge_2.13.0-M5"
```
I'd keep this one: 1 extra line of code to allow future zinc/sbt's to build the milestone releases of scala 2.13.

eed3si9n replied:
2.13.0-M2~5 should compile fine with 2.13.0-RC1 bridge.

dwijnand replied:
How sure are you?  I'd say keep this 1 line unless this fact is verified.

eed3si9n replied:
I am sure since there's been no change in _source_ between compiler-bridge_2.13.0-M5 (prior to this PR) vs compiler-bridge_2.13 (aka _2.13.0-RC1).

dwijnand commented on internal/zinc-ivy-integration/src/main/scala/sbt/internal/inc/ZincComponentCompiler.scala:
```diff
@@ -40,14 +40,12 @@ private[sbt] object ZincComponentCompiler {
+        case sc if (sc startsWith "2.11.") => "compiler-bridge_2.11"
+        case sc if (sc startsWith "2.12.") => "compiler-bridge_2.12"
+        case "2.13.0-M1"                   => "compiler-bridge_2.12"
+        case sc if (sc startsWith "2.13.") => "compiler-bridge_2.13.0-RC1"
```
You can drop the `if`.  And also call it `compiler-bridge_2.13.0`?

eed3si9n replied:
`_2.13.0-RC1` comes from `compilerBridge213/scalaBinaryVersion`. I guess I might as well fake that and go for `_2.13` on that subproject.

dwijnand replied:
Yeah, WFM.

eed3si9n commented:
```
[info] - should extract class dependencies from macro arguments *** FAILED ***
[info]   sbt.internal.inc.CompileFailed:
[info]   at sbt.internal.inc.AnalyzingCompiler.call(AnalyzingCompiler.scala:249)
[info]   at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:118)
[info]   at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
[info]   at sbt.internal.inc.AnalyzingCompiler.apply(AnalyzingCompiler.scala:74)
[info]   at sbt.internal.inc.CompilingSpecification.$anonfun$compileSrcs$3(CompilingSpecification.scala:191)
[info]   at scala.collection.TraversableLike$WithFilter.$anonfun$map$2(TraversableLike.scala:742)
[info]   at scala.collection.immutable.List.foreach(List.scala:392)
[info]   at scala.collection.TraversableLike$WithFilter.map(TraversableLike.scala:741)
[info]   at sbt.internal.inc.CompilingSpecification.$anonfun$compileSrcs$1(CompilingSpecification.scala:165)
[info]   at sbt.io.IO$.withTemporaryDirectory(IO.scala:480)
[info]   ...
Reporter completed abruptly with an exception after receiving event: TestFailed(Ordinal(0, 104),sbt.internal.inc.CompileFailed was thrown.,DependencySpecification,sbt.internal.inc.DependencySpecification,Some(sbt.internal.inc.DependencySpecification),Dependency phase should extract class dependencies from macro arguments,should extract class dependencies from macro arguments,Vector(),Some(Compilation failed),Some(2332),Some(IndentedText(- should extract class dependencies from macro arguments,should extract class dependencies from macro arguments,1)),Some(SeeStackDepthException),Some(sbt.internal.inc.DependencySpecification),None,pool-1-thread-1-ScalaTest-running-DependencySpecification,1554393311051).
java.io.NotSerializableException: sbt.util.InterfaceUtil$ConcreteProblem
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1184)
	at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1378)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1174)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
	at org.scalatest.tools.SocketReporter.apply(SocketReporter.scala:31)
	at org.scalatest.DispatchReporter$Propagator.$anonfun$run$10(DispatchReporter.scala:249)
	at org.scalatest.DispatchReporter$Propagator.$anonfun$run$10$adapted(DispatchReporter.scala:248)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.scalatest.DispatchReporter$Propagator.run(DispatchReporter.scala:248)
	at java.lang.Thread.run(Thread.java:748)
Exception in thread "Thread-59" java.io.WriteAbortedException: writing aborted; java.io.NotSerializableException: sbt.util.InterfaceUtil$ConcreteProblem
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1577)
	at java.io.ObjectInputStream.readArray(ObjectInputStream.java:1975)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1567)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2287)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2211)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2069)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1573)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2287)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2211)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2069)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1573)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2287)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2211)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2069)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1573)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:431)
	at org.scalatest.tools.Framework$ScalaTestRunner$Skeleton$1$React.react(Framework.scala:818)
	at org.scalatest.tools.Framework$ScalaTestRunner$Skeleton$1.run(Framework.scala:807)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.NotSerializableException: sbt.util.InterfaceUtil$ConcreteProblem
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1184)
	at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1378)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1174)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
	at org.scalatest.tools.SocketReporter.apply(SocketReporter.scala:31)
	at org.scalatest.DispatchReporter$Propagator.$anonfun$run$10(DispatchReporter.scala:249)
	at org.scalatest.DispatchReporter$Propagator.$anonfun$run$10$adapted(DispatchReporter.scala:248)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.scalatest.DispatchReporter$Propagator.run(DispatchReporter.scala:248)
	... 1 more
```

eed3si9n committed:
Get tests to pass on 2.13.0-RC1

Some of the extracted names are coming out very differently (mostly with extra names), so I had to comment them out.

eed3si9n committed:
call it compiler-bridge_2.13

dwijnand commented on build.sbt:
```diff
@@ -504,13 +504,13 @@ lazy val compilerBridgeTemplate: Project = (project in internalPath / "compiler-
+    //     case x if x startsWith "2.13." => ("org.scala-sbt" % "compiler-bridge_2.13.0-M2" % "1.1.0-M1-bootstrap2" % Compile).sources()
+    //     case _ => old
+    //   }
+    // },
```
delete now?

eed3si9n replied:
Yea.

eed3si9n committed:
Remove compiler bridge bootstrap code

jvican commented:
I’m reviewing this today

jvican commented on internal/compiler-bridge-test/src/test/scala/sbt/internal/inc/ExtractUsedNamesPerformanceSpecification.scala:
```diff
@@ -23,8 +26,22 @@ class ExtractUsedNamesPerformanceSpecification extends CompilingSpecification {
+    "Option",
+    "?0",
+    "package",
+    "acme",
```
Why is `acme` and `ModuleSerializationProxy` an exception for Scala 2.13?

eed3si9n replied:
acme was missing from the list of extracted names, and ModuleSerializationProxy showed up as an extracted name. We can open issue for these.

jvican commented on internal/compiler-bridge-test/src/test/scala/sbt/internal/inc/ExtractUsedNamesPerformanceSpecification.scala:
```diff
@@ -43,22 +60,26 @@ class ExtractUsedNamesPerformanceSpecification extends CompilingSpecification {
+    // val expectedNamesForTuplerInstances = Set("E", "Tuple4", "e", "case7", "Tuple15", "s", "case19", "T7", "x", "TuplerInstances", "matchEnd19", "T20", "Tuple11", "HNil", "matchEnd6", "p16", "$anon", "T19", "p20", "T2", "p10", "case22", "p19", "n", "Tuple12", "case11", "Tuple22", "p12", "matchEnd7", "N", "p4", "T13", "case26", "Tuple19", "p7", "p5", "j", "Out", "T", "p23", "case15", "matchEnd2 ...
     val expectedNamesForRefinement = Set("Out0")
-    val `expectedNamesFor::` = Set("x", "T2", "ScalaRunTime", "Iterator", "T", "head", "asInstanceOf", "Boolean", "A", "$" + "isInstanceOf", "T1", "||", "acme;::;init;", "::", "Nothing", "x$1", "any2stringadd", "acme", "typedProductIterator", "tail", "Tuple2", "AnyRef", "isInstanceOf", "Int", "java;lang;Object;init;", "_hashCode", "apply", "Object", "x$0", "==", "Some", "IndexOutOfBoundsException ...
+    // val `expectedNamesFor::` = Set("x", "T2", "ScalaRunTime", "Iterator", "T", "head", "asInstanceOf", "Boolean", "A", "$" + "isInstanceOf", "T1", "||", "acme;::;init;", "::", "Nothing", "x$1", "any2stringadd", "acme", "typedProductIterator", "tail", "Tuple2", "AnyRef", "isInstanceOf", "Int", "java;lang;Object;init;", "_hashCode", "apply", "Object", "x$0", "==", "Some", "IndexOutOfBoundsExcept ...
```
Can we compile to 2.13.0-RC1 while not disabling all of these checks? The look material to making incremental compilation work correctly in 2.13.x.

eed3si9n replied:
> The look material to making incremental compilation work correctly in 2.13.x.

I agree it does look like a problem, and I am trying to surface that by running these tests against 2.13.0-RC1.
The status quo is that the tests are running against 2.13.0-M2 or M5 and people are using the bridge source anyways to run RC1.

jvican replied:
Can we look into it? I don't feel comfortable merging this PR with all these tests disabled for all compiler bridge versions and without a deep understanding of why these failures happen.

jvican commented on internal/compiler-bridge-test/src/test/scala/sbt/internal/inc/ExtractUsedNamesSpecification.scala:
```diff
@@ -212,73 +212,76 @@ class ExtractUsedNamesSpecification extends CompilingSpecification {
-        |object Usage extends $sealedClassName
-        |object Usage2 extends $sealedClassName
-      """.stripMargin
+  // This doesn't work in 2.13.0-RC1
```
I'm fine disabling this, but a nitpick to reduce noise: can we use `/*` and `*/` instead of `//` please?

jvican replied:
@eed3si9n Before merging, can you address this nitpick? The rest of the feedback can be handled in another PR

alexarchambault commented:
~Just so you know, not publishing this prevents building 2.13.0-RC1 projects with mill, which blocks building fastparse with 2.13.0-RC1, which itself blocks building large parts of the ecosystem with 2.13.0-RC1…~

~The earlier this can be published, the better, I guess.~

(Edit: scratch that, there are workarounds actually…)

jvican commented on internal/compiler-bridge-test/src/test/scala/sbt/internal/inc/ExtractUsedNamesPerformanceSpecification.scala:
```diff
@@ -23,8 +26,22 @@ class ExtractUsedNamesPerformanceSpecification extends CompilingSpecification {
+    "package",
+    "acme",
+    "ModuleSerializationProxy",
+    "scala;runtime;ModuleSerializationProxy;init;",
```
Not tracking this name in Scala 2.13 will give incremental compiler errors when initializers are involved

jvican commented on internal/compiler-bridge-test/src/test/scala/sbt/internal/inc/ExtractUsedNamesPerformanceSpecification.scala:
```diff
@@ -23,8 +26,22 @@ class ExtractUsedNamesPerformanceSpecification extends CompilingSpecification {
+    "ModuleSerializationProxy",
+    "scala;runtime;ModuleSerializationProxy;init;",
+    "Class",
+    "Sealed"
```
This is most likely the reason why we don't track incrementality for sealed class hierarchies anymore in Scala 2.13. Not as severe correctness-wise as the other error, but still worth fixing...

eed3si9n replied:
The capitalized Class and Sealed are new named that started to show up in 2.13, but didn't exist before.
We probably should filter it out so it doesn't lead to overcompilation.

jvican commented:
@eed3si9n and I have agreed to merge this and follow up with another PR re-enabling all disabled tests, this is tracked in https://github.com/sbt/zinc/issues/653
  
It doesn't look like this will need to be merged and released directly though, Zinc users can use the bridge of Scala 2.13.0-M2 to compile Scala 2.13.0-RC1 code. Nevertheless, I'll let @eed3si9n decide what the best way to move forward is

eed3si9n committed:
Remove compiler bridge cache

eed3si9n committed:
Add scala-integration resolver
